adapters/play/src/main/scala/caliban/PlayAdapter.scala

```diff
@@ -0,0 +1,218 @@
+package caliban
+
+import akka.stream.{ Materializer, OverflowStrategy, QueueOfferResult }
+import akka.stream.scaladsl.{ Flow, Sink, Source, SourceQueueWithComplete }
+import caliban.ResponseValue.{ ObjectValue, StreamValue }
+import caliban.Value.NullValue
+import caliban.interop.play.json.parsingException
+import play.api.http.Writeable
+import play.api.libs.json.{ Json, JsValue, Writes, _ }
+import play.api.mvc.{ Action, ActionBuilder, AnyContent, PlayBodyParsers, Request, RequestHeader, Result, WebSocket }
+import play.api.mvc.Results.Ok
+import zio.{ CancelableFuture, Fiber, IO, RIO, Ref, Runtime, Schedule, Task, ZIO }
+import zio.clock.Clock
+import zio.duration.Duration
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.Try
+
+trait PlayAdapter {
+
+  def actionBuilder: ActionBuilder[Request, AnyContent]
+  def parse: PlayBodyParsers
+
+  implicit def writableGraphQLResponse[E](implicit wr: Writes[GraphQLResponse[E]]): Writeable[GraphQLResponse[E]] =
+    Writeable.writeableOf_JsValue.map(wr.writes)
+
+  private def parseJson(s: String): Try[JsValue] =
+    Try(Json.parse(s))
+
+  private def getGraphQLRequest(
+    query: String,
+    op: Option[String],
+    vars: Option[String],
+    exts: Option[String]
+  ): Either[Throwable, GraphQLRequest] = {
+    val variablesJs  = vars.flatMap(parseJson(_).toOption)
+    val extensionsJs = exts.flatMap(parseJson(_).toOption)
+    val fields = List("query" -> JsString(query)) ++
+      op.map(o => "operationName"         -> JsString(o)) ++
+      variablesJs.map(js => "variables"   -> js) ++
+      extensionsJs.map(js => "extensions" -> js)
```

Comment:
Could this just be:

```scala
Json.obj(
  "query" -> query,
  "operationName" -> op,
  "variables" -> variablesJs,
  "extensions" -> extensionsJs
).validate[GraphQLRequest]
```

Also technically, query can be empty if this is a `persisted-query`, so I think you can make `query` optional and let the interpreter handle it.

Reply:
I agree with you, `query` should be optional. I've done it already. I think we need to check this in the other adapters since there is a mismatch between them. If @ghostdogpr agrees too, I can adjust it in another PR.

Reply:
It was already supposed to be optional, and I think it's the case in http4s/uzhttp adapters. It's a bug that it's not in the Akka one 😄 If you can fix it at the same time that'd be great!

Reply:
Since this PR is ready and a release is coming soon, I'll merge it and this can be fixed in another PR as you suggested.